mlflow/entities/span.py

```diff
@@ -14,3 +40,434 @@ class SpanType:
     EMBEDDING = "EMBEDDING"
     RERANKER = "RERANKER"
     UNKNOWN = "UNKNOWN"
+
+
+class Span:
+    """
+    A span object. A span represents a unit of work or operation and is the building
+    block of Traces.
+
+    This Span class represents immutable span data that is already finished and persisted.
+    The "live" span that is being created and updated during the application runtime is
+    represented by the :py:class:`LiveSpan <mlflow.entities.LiveSpan>` subclass.
+    """
+
+    def __init__(self, otel_span: OTelReadableSpan):
+        if not isinstance(otel_span, OTelReadableSpan):
+            raise MlflowException(
+                "The `otel_span` argument for the Span class must be an instance of ReadableSpan, "
+                f"but got {type(otel_span)}.",
+                INVALID_PARAMETER_VALUE,
+            )
+
+        self._span = otel_span
+        # Since the span is immutable, we can cache the attributes to avoid the redundant
+        # deserialization of the attribute values.
+        self._attributes = _CachedSpanAttributesRegistry(otel_span)
+
+    @property
+    @lru_cache(maxsize=1)
```

Comment:
Can we use https://docs.python.org/3/library/functools.html#functools.cached_property?

Reply:
Great idea!

Reply:
@harupy Hmm it seems `cached_property` has a bit unique behavior that allows direct write to the attribute like `span.request_id = xxx`. I think we shouldn't allow such mutation.

> The mechanics of [cached_property()](https://docs.python.org/3/library/functools.html#functools.cached_property) are somewhat different from [property()](https://docs.python.org/3/library/functions.html#property). A regular property blocks attribute writes unless a setter is defined. In contrast, a cached_property allows writes.

Reply:
Got it, then never mind!